### PR TITLE
Jet PU ID scale factors

### PR DESCRIPTION
**Pull request by kondratyevd**
Started implementing jet PU ID scale factors using the recipe described here:
[https://indico.cern.ch/event/856106/contributions/3602928/attachments/1934922/3206279/puid.pdf](https://indico.cern.ch/event/856106/contributions/3602928/attachments/1934922/3206279/puid.pdf)

So far only loop over jets is implemented and all efficiencies and scale factors are set to 1. As a next step, I'll implement reading the efficiencies and scale factors from provided histograms.

### CODE REVIEW

**Comment by jpata** [comment]
Thanks for letting us know you're working on this!

Just had a quick look - as I understand, the total PU weight for an event with multiple jets is a combinatorial product of the individual PU scale factors, similar to the b-tagging reweighting: https://twiki.cern.ch/twiki/bin/view/CMS/BTagSFMethods#1a_Event_reweighting_using_scale

Is this correct? If so, consider the following numba function:

```python
@numba.njit(parallel=True)
def compute_eff_product(offsets, jets_mask_passes_id, jets_eff, out_proba):
    #loop over events in parallel
    for iev in numba.prange(len(offsets)-1):
        p_tot = 1.0
        #loop over jets in event
        for ij in range(offsets[iev], offsets[iev+1]):
            this_jet_passes = jets_mask_passes_id[ij]
            if this_jet_passes:
                p_tot *= jets_eff[ij]
            else:
                p_tot *= 1.0 - jets_eff[ij]
        out_proba[iev] = p_tot

...

jets_pu_eff, jets_pu_sf, _ = get_puid_eff_sf(calibration_file, jets.pt, jets.eta)
p_puid_mc = NUMPY_LIB.zeros(nev)
p_puid_data = NUMPY_LIB.zeros(nev)
compute_eff_product(jets.offsets, passed_puid, jets_pu_eff, p_puid_mc)
compute_eff_product(jets.offsets, passed_puid, jets_pu_eff*jets_pu_sf, p_puid_data)
eventweight_puid = p_puid_data/p_puid_mc
```

Would it perhaps be simpler & faster? Let me know if you had something else in mind!

**Comment by kondratyevd** [comment]
Hi @jpata,
yes, the PU ID weight should be implemented similarly to b-tag reweighting, so this will work. Thanks for the code - I didn't know parallel loops could be implemented - seems to be faster indeed. 

Perhaps I'll use this approach sometime in the future to rewrite FSR recovery as well.

**Comment by kondratyevd** [comment]
Hi @jpata,
I have separated reading the input files from evaluating the weights, such that the reading is done only once - I hope that is what you meant. 
Let me know if anything else should be changed in order for this implementation to fit well into the framework!

File: tests/hmm/hmumu_utils.py [hunk 1]
```diff
@@ -1520,6 +1527,61 @@ def get_selected_jets(
...
+    extractors = (ext_eff, ext_sf)
+    hist_names = (h_eff_name, h_sf_name)
+    return extractors, hist_names
+
+def jet_puid_evaluate(extractors, hist_names, jet_pt, jet_eta):
+    ext_eff, ext_sf = extractors
+    h_eff_name, h_sf_name = hist_names
+
+    ext_eff.finalize()
+    evaluator_eff = ext_eff.make_evaluator()
```
Comment (jpata) [inline tests/hmm/hmumu_utils.py]: Very good!

I think you should also be able to move the `finalize()` and `make_evaluator()` to the function `jet_puid_extract()`, such that this function becomes `jet_puid_evalute(puid_evaluator, jet_pt, jet_eta)` and already takes just the object `puid_evaluator=evaluator_eff[h_eff_name]` as one of the arguments.

This way you don't have to redo the finalization every time you wan to compute the puid values.

See e.g. https://github.com/jpata/hepaccelerate-cms/blob/master/tests/hmm/analysis_hmumu.py#L103 how the JES stuff is finalized once in the beginning of the code.

**Comment by jpata** [comment]
Thanks - looks good! I added a small comment inline, let me know if that sounds reasonable to you. Otherwise, I would say it's good to go. Perhaps @nanlu06 has any other comments?

**Comment by nanlu06** [comment]
Hi Dmitry! Thank you very much. Very nice! I think we should merge this.

**Comment by jpata** [comment]
Thanks @nanlu06 for taking a look and @kondratyevd for the work! I will merge this in a few hours, unless there are additional updates.

**Pull request merged**